site/documentation/content/api/device-connection/index.md

```diff
@@ -126,6 +126,7 @@ The following table provides an overview of the properties a client needs to set
 | :-------------------- | :-------: | :----------------------- | :-------- | :---------- |
 | *subject*             | yes       | *properties*             | *string*  | MUST be set to `set-cmd-handling-adapter-instance`. |
 | *adapter_instance_id* | yes       | *application-properties* | *string*  | The identifier of the protocol adapter instance that currently handles commands for the device or gateway identified by the *device_id* property. |
+| *lifespan*            | no        | *application-properties* | *int*     | The lifetime of the mapping entry in seconds. After that period, the mapping entry shall be treated as non-existent by the *Device Registration API* methods. A negative value, as well as an omitted property, is interpreted as an unlimited lifespan.|
```

Comment:
what if the client provides the value but the implementation doesn't support it?

Reply:
I've adapted the description as follows:
````
The desired lifespan of the mapping entry in seconds. If the *Device Registration API* implementation supports this property, the mapping entry shall be treated as non-existent after that period has elapsed. A negative value, as well as an omitted property, is interpreted as an unlimited lifespan. A *Device Registration API* implementation may not support this property, meaning that a client can't rely on the given lifespan to be applied.
````

Reply:
@sophokles73 An afterthought on this:
Since we target this change for the next minor version, meaning we can introduce incompatible changes to _providers_ of the Device Connection API, I think we _can_ indeed _require_ implementations of the API to support the `lifespan` property. And I think we should make it a requirement then (otherwise we have to maintain extra client logic for the case it isn't supported by an implementation). WDYT?

Reply:
How could a client know if an implementation supports it or not? The implementation would simply ignore the given value and keep the mapping until it evicts it, wouldn't it?

Reply:
My point is to make it _mandatory_ for an implementation to support the `lifespan` property, so that a client can be sure it is supported.

Reply:
I understand that. My note was in reply to your remark
> otherwise we have to maintain extra client logic for the case it isn't supported by an implementation

We have no way of making the registry supporting the property because the adapter has no way of detecting this and thus cannot e.g. issue a warning or shut down or use additional logic to compensate for it. I agree, though, that we should document the fact that, if a registry does not support the additional lifespan property, then the adapter might fail to correctly route commands.

Reply:
And as a general note concerning the `Mandatory` column in the `Request Message Format` properties table: 
I understand that in such a way that it is or isn't mandatory for the _client_ to use the property in its requests. If not stated otherwise in the description, I would think that API _implementations_ _have_ to support the property regardless, right?

Reply:
> We have no way of making the registry supporting the property because the adapter has no way of detecting this and thus cannot e.g. issue a warning or shut down or use additional logic to compensate for it.

Well, we could use the existing logic of always explicitly removing the mapping entry (with the notable exception of edge cases like when the adapter is killed before being able to do that), noticing in the return value, whether the (expired) entry was actually removed before by the API implementation.
But I would rather omit this and make it part of the contract that the API implementation just has to support the `lifespan` property.

> I agree, though, that we should document the fact that, if a registry does not support the additional lifespan property, then the adapter might fail to correctly route commands.

Yes, we can insert this as the reason why it has to be implemented in the API implementation.

Reply:
> Since we target this change for the next minor version, meaning we can introduce incompatible changes to providers of the Device Connection API, I think we can indeed require implementations of the API to support the lifespan property.

I am not sure about this, because the (public) Device Connection API describes a dependency on an external system as opposed to APIs where Hono is the actual implementation (provider) of the API, e.g. the API defined by its protocol adapters towards the devices.

@dejanb @ctron WDYT?

Reply:
Created PR #1917 for adapting the documentation.

Reply:
> I am not sure about this, because the (public) Device Connection API describes a dependency on an external system as opposed to APIs where Hono is the actual implementation (provider) of the API, e.g. the API defined by its protocol adapters towards the devices.

I was sticking to the OSGi semantic versioning notions of _providers_ and _consumers_ (with the emphasis of keeping only _consumers_ stable between minor versions) here and seeing the device registry as a _provider_ of the different Hono APIs.

But, yes, in the sense that the device registry is external, it could be seen as even more important to keep things compatible for it between Hono minor versions than for the protocol adapter _consumers_, which are Hono internal (in a way).

But I think, if we go that way of keeping device registry implementations compatible between Hono minor versions, we have to increase Hono major versions in more cases. For example, the Command & Control changes in 1.2 (with its additional, non-optional Device Connection API methods) would have required a 2.0.
And now, the addition of another additional API flag, mentioned in https://github.com/eclipse/hono/issues/1858#issuecomment-617630849, where the support in the implementation should better be not optional, would strictly speaking require us to go for 2.0 instead of 1.3 as well.
In any case, we should decide on which approach to take.

Reply:
The registry APIs defined as part of Hono all define APIs that Hono *consumes*, i.e. they define what Hono expects to find implemented by some third party entity. The fact that we also provide an example implementation of these APIs doesn't mean that we can change these APIs in a non-backward compatible way as long as we implement these changes in the example registry. So, indeed, I agree that we already would have needed to increase the version to 2.0.0 instead of 1.2.0 because of the newly introduced, mandatory operation in the Device Connection API.

@ctron @dejanb WDYT?

Reply:
On the community call last Thursday it was decided that we want to stick to the OSGi semantic versioning approach.
An implementation of the device registry APIs is seen as a _provider_. Hence there can be incompatible changes between Hono _minor_ versions for the _providers_, requiring changes to the implementation of the device registry APIs. But, at the same time, these API changes must keep the _consumers_, i.e. the protocol adapter implementations, compatible with the previous minor version.
That means that a device registry implementation of a newer minor version still is compatible with protocol adapters of an older minor version of the Hono APIs.
This was seen as the key point here.
It is the obligation of the device registry implementation, to keep up with additive changes in newer minor versions of the APIs then.